nifi-nar-bundles/nifi-pgp-bundle/nifi-pgp-processors/src/main/java/org/apache/nifi/processors/pgp/DecryptContentPGP.java

```diff
@@ -0,0 +1,386 @@
+                    literalData = (PGPLiteralData) object;
+                    break;
+                }
+            }
+
+            if (literalData == null) {
+                throw new PGPProcessException("PGP Literal Data not found");
+            }
+
+            return literalData;
+        }
+
+        private InputStream getDecryptedDataStream(final PGPEncryptedData encryptedData) throws PGPException {
+            getLogger().debug("PGP Encrypted Data [{}] Found", getEncryptedDataType(encryptedData));
+
+            if (encryptedData instanceof PGPPBEEncryptedData) {
+                return getDecryptedDataStream((PGPPBEEncryptedData) encryptedData);
+            } else if (encryptedData instanceof PGPPublicKeyEncryptedData) {
+                return getDecryptedDataStream((PGPPublicKeyEncryptedData) encryptedData);
+            } else {
+                final String message = String.format("PGP Encrypted Data [%s] Not Supported", getEncryptedDataType(encryptedData));
+                throw new UnsupportedOperationException(message);
+            }
+        }
+
+        private InputStream getDecryptedDataStream(final PGPPBEEncryptedData passwordBasedEncryptedData) throws PGPException {
+            if (passphrase == null) {
+                throw new PGPProcessException("PGP Password-Based Encryption Found: Passphrase not configured");
+            } else {
+                final PBEDataDecryptorFactory decryptorFactory = new BcPBEDataDecryptorFactory(passphrase, new BcPGPDigestCalculatorProvider());
+                final int symmetricAlgorithm = passwordBasedEncryptedData.getSymmetricAlgorithm(decryptorFactory);
+                attributes.put(PGPAttributeKey.SYMMETRIC_KEY_ALGORITHM_ID, Integer.toString(symmetricAlgorithm));
+                return passwordBasedEncryptedData.getDataStream(decryptorFactory);
+            }
+        }
+
+        private InputStream getDecryptedDataStream(final PGPPublicKeyEncryptedData publicKeyEncryptedData) throws PGPException {
+            if (privateKeyService == null) {
+                throw new PGPProcessException("PGP Public Key Encryption Found: Private Key Service not configured");
+            } else {
+                final long keyId = publicKeyEncryptedData.getKeyID();
+                final Optional<PGPPrivateKey> foundPrivateKey = privateKeyService.findPrivateKey(keyId);
+                if (foundPrivateKey.isPresent()) {
+                    final PGPPrivateKey privateKey = foundPrivateKey.get();
+                    final PublicKeyDataDecryptorFactory decryptorFactory = new BcPublicKeyDataDecryptorFactory(privateKey);
+                    final int symmetricAlgorithm = publicKeyEncryptedData.getSymmetricAlgorithm(decryptorFactory);
+                    attributes.put(PGPAttributeKey.SYMMETRIC_KEY_ALGORITHM_ID, Integer.toString(symmetricAlgorithm));
+                    return publicKeyEncryptedData.getDataStream(decryptorFactory);
```

Comment:
One note here:
-  When I specify only valid Public/Private Key services in the EncryptContentPGP and DecryptContentPGP processors but no passphrases, both encryption and decryption work.  Likewise, when I specify only a matching passphrase in both processors, both encryption and decryption work.
- However, when I use the above working configuration of Public/Private Key services, and then add the same matching passphrase to both processors, I get the following error in DecryptContentPGP:
`2021-04-02 16:09:17,767 ERROR [Timer-Driven Process Thread-8] o.a.n.processors.pgp.DecryptContentPGP DecryptContentPGP[id=93e3a4e0-0178-1000-a8d0-00468239c9f4] Decryption Failed StandardFlowFileRecord[uuid=9675b046-b929-4b38-81fa-e3e20f8149e8,claim=StandardContentClaim [resourceClaim=StandardResourceClaim[id=1617385030974-1, container=default, section=1], offset=17471, length=672],offset=0,name=9675b046-b929-4b38-81fa-e3e20f8149e8,size=672]: org.apache.nifi.processors.pgp.exception.PGPDecryptionException: PGP Decryption Failed [Public Key Encryption]
org.apache.nifi.processors.pgp.exception.PGPDecryptionException: PGP Decryption Failed [Public Key Encryption]
	at org.apache.nifi.processors.pgp.DecryptContentPGP$DecryptStreamCallback.getLiteralData(DecryptContentPGP.java:262)
	at org.apache.nifi.processors.pgp.DecryptContentPGP$DecryptStreamCallback.process(DecryptContentPGP.java:238)
	at org.apache.nifi.controller.repository.StandardProcessSession.write(StandardProcessSession.java:3054)
	at org.apache.nifi.processors.pgp.DecryptContentPGP.onTrigger(DecryptContentPGP.java:159)
	at org.apache.nifi.processor.AbstractProcessor.onTrigger(AbstractProcessor.java:27)
	at org.apache.nifi.controller.StandardProcessorNode.onTrigger(StandardProcessorNode.java:1180)
	at org.apache.nifi.controller.tasks.ConnectableTask.invoke(ConnectableTask.java:214)
	at org.apache.nifi.controller.scheduling.TimerDrivenSchedulingAgent$1.run(TimerDrivenSchedulingAgent.java:103)
	at org.apache.nifi.engine.FlowEngine$2.run(FlowEngine.java:110)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.bouncycastle.openpgp.PGPException: Exception starting decryption
	at org.bouncycastle.openpgp.PGPPublicKeyEncryptedData.getDataStream(Unknown Source)
	at org.apache.nifi.processors.pgp.DecryptContentPGP$DecryptStreamCallback.getDecryptedDataStream(DecryptContentPGP.java:324)
	at org.apache.nifi.processors.pgp.DecryptContentPGP$DecryptStreamCallback.getDecryptedDataStream(DecryptContentPGP.java:295)
	at org.apache.nifi.processors.pgp.DecryptContentPGP$DecryptStreamCallback.getLiteralData(DecryptContentPGP.java:257)
	... 15 common frames omitted
Caused by: java.io.EOFException: null
	at org.bouncycastle.openpgp.PGPEncryptedData$TruncatedStream.<init>(Unknown Source)
	... 19 common frames omitted`

Reply:
Although this is somewhat of an edge case, it is a great example of a message containing multiple Encrypted Data Packets as described in RFC 4880 Section 5.3.  This uncovered an issue with the handling of multiple Encrypted Data elements in the callback.  The `DecryptContentPGP` processor should support decryption using either a Public Key or Passphrase, depending on property configuration and data received, so I will push an update to correct data handling in this scenario.